src/main/scala/org/scalacheck/Gen.scala

```diff
@@ -631,11 +624,29 @@ object Gen extends GenArities with GenVersionSpecific {
    *  complete container generator will also fail. */
   def buildableOfN[C,T](n: Int, g: Gen[T])(implicit
     evb: Buildable[T,C], evt: C => Traversable[T]
-  ): Gen[C] =
-    sequence[C,T](Traversable.fill(n)(g)) suchThat { c =>
-      // TODO: Can we guarantee c.size == n (See issue #89)?
-      evt(c).forall(g.sieveCopy)
+  ): Gen[C] = {
+    require(n >= 0, s"invalid size given: $n")
+    gen { (p, seed0) =>
+      var seed: Seed = seed0
+      val bldr = evb.builder
+      val allowedFailures = Integer.max(10, n / 10)
```

Comment:
kind of a bummer that this totally ad-hoc. I wonder if it should be in `p`? Maybe as some kind of filter/retry strategy? Might require more thought, but maybe add a `TODO`. It would be interesting to think about filter failure strategies in context of `flatMap/zip`. Large flatmaps/zips cause exponential decrease in filter success rates if you retry the whole chain. But if you retry after each step, you can keep the success rate constant.

e.g.: if you zip N things, each successful with probability `p`, the full zip has success `p^N`, so, the expected number of retries to get a single item: `(1/p)^N = exp(N ln(1/p))`. But if you retry each step in the chain `K` times for a success, you get `p_step = 1 - (1 - p)^K`. and `1/p_step^N = 1/(1 - (1 - p)^K))^N ~ exp(N(1 - p)^K)` If you choose `K ~ log N` then `exp(N (1 - p)^K) ~ exp(N exp(-pK)) = exp(N (1 / N) ^ p) ~ 1`

So, what this means is if you have a zip chain `N` long, and you retry each step `O(log N)` times, you get `O(1)` total success rate, which means we only do `N log N` work.

For flatMap, we don't know how deep we are going to go, indeed, it can be dynamic, but for zip we statically know, so we could just do `log_2(Zip width)` size retries.

By the way, this analysis suggests if you know you have `N` items, like we do in this combinator, we want to retry `O(log N)` not `c N` times. I think  with linear retries, which you have here, you should make failure of the entire generator exponentially low. Maybe linear is actually fine, since you have to do linear work already, the total work here is still linear.

Reply:
Yeah I would love to be threading through something like a total retries parameter and then a used retries count. In the absence of that I've sort of erred on the side of minimizing filtering in the library itself while also adding retries.

I'm a bit nervous about adding retrying to flatMap/zip in this PR, so I'd rather wait and think about how best to do that. I do like the idea you had for zip of apportioning the retries across all the underlying generators, for example.